### PR TITLE
Remove GTF_GLOB_REF on a few more addressing nodes

### LINKED ISSUE #68669: Assertion failed 'node->DefinesLocalAddr(this, size, &lcl, nullptr) && lvaGetDesc(lcl)->lvHiddenBufferStructArg'
pipeline: runtime-coreclr libraries-jitstress
Test: System.Reflection.MetadataLoadContext.Tests
JitStress=1
Failed for both win-x64 and linux-arm

https://dev.azure.com/dnceng/public/_build/results?buildId=1741924&view=ms.vss-test-web.build-test-results-tab&runId=47086172&resultId=193772&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

```
+ printenv
+ grep COMPlus
COMPlus_TieredCompilation=0
COMPlus_JitStress=1
COMPlus_DbgMiniDumpName=/home/helixbot/dotnetbuild/dumps/coredump.%d.dmp
COMPlus_DbgEnableMiniDump=1
+ ./RunTests.sh --runtime-path /root/helix/work/correlation
----- start Thu Apr 28 07:47:43 UTC 2022 =============== To repro directly: =====================================================
pushd .
/root/helix/work/correlation/dotnet exec --runtimeconfig System.Reflection.MetadataLoadContext.Tests.runtimeconfig.json --depsfile System.Reflection.MetadataLoadContext.Tests.deps.json xunit.console.dll System.Reflection.MetadataLoadContext.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing 
popd
===========================================================================================================
/root/helix/work/workitem/e /root/helix/work/workitem/e
  Discovering: System.Reflection.MetadataLoadContext.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Reflection.MetadataLoadContext.Tests (found 493 of 494 test cases)
  Starting:    System.Reflection.MetadataLoadContext.Tests (parallel test collections = on, max threads = 4)

Assert failure(PID 61 [0x0000003d], Thread: 75 [0x004b]): Assertion failed 'node->DefinesLocalAddr(this, size, &lcl, nullptr) && lvaGetDesc(lcl)->lvHiddenBufferStructArg' in 'System.Reflection.TypeLoading.Ecma.EcmaEvent:ComputeEventRaiseMethod():System.Reflection.TypeLoading.RoMethod:this' during 'Morph - Global' (IL size 37; hash 0x941ee672; FullOpts)

    File: /__w/1/s/src/coreclr/jit/gentree.cpp Line: 17913
    Image: /root/helix/work/correlation/dotnet
```

@dotnet/jit-contrib

### PR DESCRIPTION
**Pull request by jakobbotsch**
GTF_GLOB_REF is necessary on a GT_ADDR node only if the child actually
uses a tree with this flag as a value (and not as a location). This
removes the flag in a few more places by using IsLocalAddrExpr.

Fixes #68669

I've also improved the assertion for not breaking recognition of the retbuf.
I realized that in release builds even if we did introduce a temp, the noway_assert that I added would pass.

### CODE REVIEW

File: src/coreclr/jit/fgdiagnostic.cpp [hunk 1]
```diff
@@ -3085,8 +3085,8 @@ void Compiler::fgDebugCheckFlags(GenTree* tree)
         return GenTree::VisitResult::Continue;
     });
 
-    // ADDR nodes break the "parent flags >= operands flags" invariant for GTF_GLOB_REF.
-    if (tree->OperIs(GT_ADDR) && op1->OperIs(GT_LCL_VAR, GT_LCL_FLD))
+    // Addresses of locals never need GTF_GLOB_REF
```
Comment (SingleAccretion) [inline src/coreclr/jit/fgdiagnostic.cpp:3088]: Nit: could you keep the mention of the `ADDR` in the comment (here and in morph)?

It makes it stand out as the quirk this special handling is (and will help find places that need to be updated when `ADDR` is deleted).

Comment (jakobbotsch) [reply src/coreclr/jit/fgdiagnostic.cpp:3088]: I've readded the check for GT_ADDR, I don't believe we need to handle other cases anyway as they will just be handled when it's their turn in morph anyway.

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @kunalspathak, this is a fix for #68669

**Review by kunalspathak — CHANGES_REQUESTED** [review]
I realize that we should also check for `!varDsc->IsAddressExposed()` below:

https://github.com/dotnet/runtime/blob/6cc626f1d714f50f4ff5dd0bedf3d1ae71bbdb19/src/coreclr/jit/lclmorph.cpp#L687

Looking at the tree that you have in the repro, while visiting `[000196]` , we mark it as address-exposed and then while visiting `[000198]` we mark it as hidden buffer.

```
LocalAddressVisitor visiting statement:
STMT00044 ( INL14 @ 0x015[E-] ... ??? ) <- INL12 @ ??? <- INL09 @ 0x00B[E-] <- INLRT @ ???
               [000198] S-C-G-------                        ▌  CALL      void   System.Reflection.TypeLoading.Ecma.MetadataExtensions.GetMethodDefinition
               [000203] ------------ retbuf                 ├──▌  ADDR      byref 
               [000202] -------N----                        │  └──▌  FIELD     struct hackishFieldName
               [000191] ------------                        │     └──▌  ADDR      byref 
               [000192] -------N----                        │        └──▌  LCL_VAR   struct<System.Reflection.TypeLoading.Ecma.EcmaMethodDecoder, 32> V16 tmp14        
               [000200] n----------- arg1                   ├──▌  OBJ       struct<System.Reflection.Metadata.MethodDefinitionHandle, 4>
               [000199] ------------                        │  └──▌  ADDR      byref 
               [000193] -------N----                        │     └──▌  LCL_VAR   struct<System.Reflection.Metadata.MethodDefinitionHandle, 4> V18 tmp16        
               [000196] --C-G------- arg2                   └──▌  CALL      ref    System.Reflection.TypeLoading.Ecma.EcmaMethodDecoder.get_Reader
               [000194] ------------ this                      └──▌  ADDR      byref 
               [000195] -------N----                              └──▌  LCL_VAR   struct<System.Reflection.TypeLoading.Ecma.EcmaMethodDecoder, 32> V16 tmp14        

Local V16 should not be enregistered because: it is address exposed

Local V16 should not be enregistered because: it is hidden buffer struct arg
```

**Comment by SingleAccretion** [comment]
What's the reason for the special handling for address-exposed locals?

(Noting that previously non-exposed locals can become exposed "later on", as late as morph in fact)

**Comment by kunalspathak** [comment]
> (Noting that previously non-exposed locals can become exposed "later on", as late as morph in fact)

Here we explicitly not mark something address exposed if something was marked as "hidden buffer", so probably we should avoid marking something "hidden buffer" if it is addressed exposed. But yes, it can be marked exposed later on.

**Comment by SingleAccretion** [comment]
> Here we explicitly not mark something address exposed if something was marked as "hidden buffer", so probably we should avoid marking something "hidden buffer" if it is addressed exposed.

Well, the reason we don't mark it exposed is because we can reason about the liveness now.

I suppose the reason for not marking it `(RB)` would would be to avoid looking at these locals down the line (i. e. save a little throughput).

**Comment by jakobbotsch** [comment]
> I realize that we should also check for `!varDsc->IsAddressExposed()` below:

It seems weird to me to do that, there is no guarantee that we visited the address-exposing tree before we visit the call here. So this can never be done for correctness, as @SingleAccretion points out above.

**Comment by kunalspathak** [comment]
> there is no guarantee that we visited the address-exposing tree before we visit the call here.

Agree

**Review by kunalspathak — APPROVED** [review]
LGTM

**Pull request merged**